pkg/performance/ringbuffer/ringbuffer.go

```diff
@@ -0,0 +1,91 @@
+// Copyright Antimetal, Inc. All rights reserved.
+//
+// Use of this source code is governed by a source available license that can be found in the
+// LICENSE file or at:
+// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt
+
+package ringbuffer
+
+import "fmt"
+
+// RingBuffer is a generic, thread-unsafe circular buffer implementation that
+// overwrites oldest elements when capacity is reached.
+//
+// This implementation is useful for scenarios where you want to keep only
+// the most recent N items, such as:
+//   - Recent log entries
+//   - Latest measurements or samples
+//   - Rolling window of events
+//
+// Note: This implementation is NOT thread-safe. If concurrent access is needed,
+// synchronization must be handled externally.
+type RingBuffer[T any] struct {
+	data     []T
+	head     int // next write position
+	size     int // current number of elements
+	capacity int // maximum capacity
+}
+
+// New creates a new ring buffer with the given capacity
+func New[T any](capacity int) (*RingBuffer[T], error) {
+	if capacity <= 0 {
+		return nil, fmt.Errorf("capacity must be greater than 0, got %d", capacity)
+	}
+	return &RingBuffer[T]{
+		data:     make([]T, capacity),
+		capacity: capacity,
```

Comment:
note: you don't _really_ need to store capacity. You could save yourself the extra 8 bytes and use `cap(r.data)` because the slice struct will already contain the capacity field and you are never at any point increasing the capacity of the buffer